### PR TITLE
加入transformer

### PR DESCRIPTION
**Pull request by W1Fl**

### CODE REVIEW

**Comment by MorvanZhou** [comment]
有几个建议，
1. L176 v2i, i2v 在定义网络的时候最好将数据和网络定义分离开，因为数据不属于网络。
2. L184 我不太确定这种用法是不是官方建议的，有没有官方的例子吗？
3. L277 这里只定义了loss，没有调用。
4. 全局变量的命名方式不统一，大小写问题，需要修改
5. L203 这个用正负号做mask的方法很赞
6. L206 我在主分支中修改了 def translate()功能，建议你也更改一下。
7. L275 load_data()的功能只调用了一次，可以合并到train里面。
8. L72 L110 LayerNorm 有些trick，建议把默认的 axis=-1 明着写上去，并且注释一些原因
9. L30 各种w的定义可能有问题，我不太确定L38这样的点积方式是否正确，需要解释一下。比如你的 wq [h, f, f], 但实际上的wq应该是 [f, h*h_dim]
10. L44 attention 的维度可能有误，这个受上面步骤的影响

一些未来优化的考虑：
1. L268 这里一次性处理的所有数据，如果数据量大，全部有model.fit()放入网络, 内存可能会有问题，以后的项目要考虑这点

**Comment by W1Fl** [comment]
解释一下：
2. build方法的super调用我习惯写在后面，不知道会不会有影响
3. loss确实调用了，在编译的时候
7. 其实回调函数中也应该用load_data方法，写的时候忘记了
9. 我把多头作为一个新的维度来合并计算，这种写法其实是等价的，但是更直观也更高效（只需要考虑一个头的计算过程，其他类似）
10. 我交换了k和q的位置来做矩阵乘法

关于fit，这确实不是一个效率高的写法，但是很简单

**Comment by W1Fl** [comment]
额看到那个loss了，忘记删。。。

**Comment by W1Fl** [comment]
已经修复这些问题重新提交了。变量和维度太多，加上我太粗心给弄晕了
